scipy/ndimage/interpolation.py

```diff
@@ -677,76 +689,53 @@ def rotate(input, angle, axes=(1, 0), reshape=True, output=None, order=3,
-                            [-m12, m22]], dtype=numpy.float64)
-        minc = [0, 0]
-        maxc = [0, 0]
-        coor = numpy.dot(mtrx, [0, ix])
-        minc, maxc = _minmax(coor, minc, maxc)
-        coor = numpy.dot(mtrx, [iy, 0])
-        minc, maxc = _minmax(coor, minc, maxc)
-        coor = numpy.dot(mtrx, [iy, ix])
-        minc, maxc = _minmax(coor, minc, maxc)
-        oy = int(maxc[0] - minc[0] + 0.5)
-        ox = int(maxc[1] - minc[1] + 0.5)
+        iy, ix = in_plane_shape
+        out_bounds = rot_matrix.dot([[0, 0, iy, iy],
+                                     [0, ix, 0, ix]])
+        out_plane_shape = (out_bounds.ptp(axis=1) + 0.5).astype(int)
     else:
-        oy = input.shape[axes[0]]
-        ox = input.shape[axes[1]]
-    offset = numpy.zeros((2,), dtype=numpy.float64)
-    offset[0] = float(oy) / 2.0 - 0.5
-    offset[1] = float(ox) / 2.0 - 0.5
-    offset = numpy.dot(matrix, offset)
-    tmp = numpy.zeros((2,), dtype=numpy.float64)
-    tmp[0] = float(iy) / 2.0 - 0.5
-    tmp[1] = float(ix) / 2.0 - 0.5
-    offset = tmp - offset
-    output_shape = list(input.shape)
-    output_shape[axes[0]] = oy
-    output_shape[axes[1]] = ox
+        out_plane_shape = img_shape[axes]
+
+    out_center = rot_matrix.dot((out_plane_shape - 1.0) * 0.5)
+    in_center = (in_plane_shape - 1.0) * 0.5
+
+    offset = in_center - out_center
+
+    output_shape = img_shape
+    output_shape[axes] = out_plane_shape
     output_shape = tuple(output_shape)
-    output = _ni_support._get_output(output, input,
-                                     shape=output_shape)
-    if input.ndim <= 2:
-        affine_transform(input, matrix, offset, output_shape, output,
+
+    output = _ni_support._get_output(output, input, shape=output_shape)
+
+    if rank <= 2:
+        affine_transform(input, rot_matrix, offset, output_shape, output,
                          order, mode, cval, prefilter)
     else:
-        coordinates = []
-        size = numpy.prod(input.shape, axis=0)
-        size //= input.shape[axes[0]]
-        size //= input.shape[axes[1]]
-        for ii in range(input.ndim):
-            if ii not in axes:
-                coordinates.append(0)
-            else:
-                coordinates.append(slice(None, None, None))
-        iter_axes = list(range(input.ndim))
-        iter_axes.reverse()
-        iter_axes.remove(axes[0])
-        iter_axes.remove(axes[1])
-        os = (output_shape[axes[0]], output_shape[axes[1]])
-        for ii in range(size):
-            ia = input[tuple(coordinates)]
-            oa = output[tuple(coordinates)]
-            affine_transform(ia, matrix, offset, os, oa, order, mode,
-                             cval, prefilter)
-            for jj in iter_axes:
-                if coordinates[jj] < input.shape[jj] - 1:
-                    coordinates[jj] += 1
-                    break
-                else:
-                    coordinates[jj] = 0
+        coordinates_list = itertools.product(
```

Comment:
Nice code golfing but desperately needs some commenting. Rank is already confusing me since it is actually the ndim of the array and not the rank of the image array.

Reply:
I agree, I am not happy with the `rank` variable name, will change it to `ndim`.
For sure, comment is welcome here. what about: `# If ndim > 2,  the rotation is applied over all the planes parallel to axes.`?